algo/src/main/java/org/neo4j/graphalgo/impl/HugeLabelPropagation.java

```diff
@@ -0,0 +1,240 @@
+        void setExistingLabels() {
+            PrimitiveLongIterator iterator = nodes.iterator();
+            while (iterator.hasNext()) {
+                long nodeId = iterator.next();
+                long existingLabel = (long) nodeProperties.nodeWeight(nodeId, (double) nodeId);
+                existingLabels.set(nodeId, existingLabel);
+            }
+        }
+
+        @Override
+        Computation computeStep() {
+            return new ComputeStep(
+                    graph.concurrentCopy(),
+                    graph,
+                    nodeWeights,
+                    progressLogger,
+                    direction,
+                    graph.nodeCount() - 1L,
+                    existingLabels,
+                    nodes,
+                    randomizeOrder
+            );
+        }
+    }
+
+    private static final class ComputeStep extends Computation implements HugeRelationshipConsumer {
+
+        private final HugeRelationshipIterator graph;
+        private final HugeRelationshipWeights relationshipWeights;
+        private final HugeWeightMapping nodeWeights;
+        private final ProgressLogger progressLogger;
+        private final Direction direction;
+        private final long maxNode;
+        private final HugeLongArray existingLabels;
+        private final PrimitiveLongIterable nodes;
+        private final LongDoubleHashMap votes;
+
+        private ComputeStep(
+                HugeRelationshipIterator graph,
+                HugeRelationshipWeights relationshipWeights,
+                HugeWeightMapping nodeWeights,
+                ProgressLogger progressLogger,
+                Direction direction,
+                final long maxNode,
+                HugeLongArray existingLabels,
+                PrimitiveLongIterable nodes,
+                boolean randomizeOrder) {
+            this.graph = graph;
+            this.relationshipWeights = relationshipWeights;
+            this.nodeWeights = nodeWeights;
+            this.progressLogger = progressLogger;
+            this.direction = direction;
+            this.maxNode = maxNode;
+            this.existingLabels = existingLabels;
+            this.nodes = RandomlySwitchingLongIterable.of(randomizeOrder, nodes);
+            this.votes = new LongDoubleScatterMap();
+        }
+
+        @Override
+        boolean computeAll() {
+            PrimitiveLongIterator iterator = nodes.iterator();
+            boolean didChange = false;
+            while (iterator.hasNext()) {
+                didChange = compute(iterator.next(), didChange);
+            }
+            return didChange;
+        }
+
+        private boolean compute(long nodeId, boolean didChange) {
+            votes.clear();
+            long partition = existingLabels.get(nodeId);
+            long previous = partition;
+            graph.forEachRelationship(nodeId, direction, this);
+            double weight = Double.NEGATIVE_INFINITY;
+            for (LongDoubleCursor vote : votes) {
```

Comment:
is there a better way to compute the max, perhaps during the computation?
Feels like double work somehow.

Reply:
I'm not quite sure how, the votes are aggregations on the partition, we can't look for the max vote until we have gathered all votes, at least not with 100% certainty. I could employ a Heavy–Hitter/Count–min–sketch type approximation (looks like http://theory.stanford.edu/~tim/s17/l/l2.pdf could be relevant), but those have an inherent uncertainty. We might get the correct winner or might occasionally choose the wrong one and would have no way of knowing, but we can do the voting in a single pass with less memory.